ChangeLog.d/psa_allow_tweaking_library_configuration.txt

```diff
@@ -0,0 +1,10 @@
+Features
+   * The PSA crypto subsystem can now be configured to use less static RAM by
+     tweaking the settings for maximum key size and maximum amount of keys
+     simultaneously in RAM. These settings have sensible defaults when not
+     overridden explicitly.
+     MBEDTLS_PSA_KEY_SLOT_COUNT sets the maximum number of volatile keys that
+     can exist simultaneously.
+     MBEDTLS_PSA_MAX_KEY_BITS sets the maximum supported number of bits in a key
+     and has thus an impact on the sizing of potentially statically allocated
+     buffers in the storage implementation.
```

Comment:
Where does reducing `MBEDTLS_PSA_MAX_KEY_BITS` reduce the amount of RAM? I don't think we ever allocate memory based on it, and if we did that would be a bug. For a while, `PSA_MAX_KEY_BITS` was effectively `SIZE_MAX/8`.

Reducing `MBEDTLS_PSA_MAX_KEY_BITS` might be useful in a multi-application system to limit the size of key objects that any application can create, but even there I would expect that a per-application quota would be more useful than a per-key limit, which doesn't even directly control the amount of memory (a 4096-bit RSA key consumes a lot more memory than a 4096-bit key derivation key).

Reply:
I believe it was used for allocation at some point (which was the reason why we patched it in the first place), but you're right, that seems to be gone now.

There are other good reasons for having it tweakable though:
* There's a multitude of test suites testing against boundary conditions of `PSA_MAX_KEY_BITS`. For example, a key import of `PSA_MAX_KEY_BITS` is expected to succeed, whereas `PSA_MAX_KEY_BITS+1` should fail. On targets where memory allocation is done by a buffer allocator (e.g. due to functional safety requirements), instead of proceeding to try the allocation which would already be pointless, the code could return much earlier by knowing that the target cannot support such a large key. For on-target testing, I think this would be a better approach then to allow 'out of memory' errors on arbitrary test cases deemed 'large'.
* Persistent import (or really, everything to do with ITS) testing assumes an ITS backend is able to store keys as large as `PSA_MAX_KEY_BITS` (which might even be a bug, since `PSA_CRYPTO_MAX_STORAGE_SIZE` doesn't seem to take the storage format overhead into account). `PSA_CRYPTO_MAX_STORAGE_SIZE` is a function of `PSA_MAX_KEY_BITS`. So it makes sense to have `PSA_MAX_KEY_BITS` tweakable to:
    * Match the ITS backend's capabilities (small embedded targets don't often have full-fledged file systems)
    * Keep control of the target memory consumption (a small embedded target with an external file system will have to allocate a read buffer in function of the largest-possible key that is stored there)
    * Avoid the ITS backend from having to deal with oversized memory allocations

Reply:
Having thought it over, I maintain my opposition to making `MAX_KEY_BITS` a configurable API. Having a way to limit the size of an individual object does make sense if you need to allocate a static-sized buffer for the key representation, but then it should be based on the representation. In practice, for currently supported key type, this means you'd typically set a limit for raw keys and one for RSA.

The design of the interface for configuring needs more thought, so let's make it a separate concern. Please remove it from this pull request, let's talk the design through some more and make a separate pull request for it.

Reply:
I've removed all changes to MAX_KEY_BITS through a rebase.

Reply:
I filed https://github.com/ARMmbed/mbedtls/issues/4149 to continue discussing the design of `MBEDTLS_PSA_MAXIMUM_KEY_SIZE` (or whatever it ends up being called).